src/ldapclient_lib.py

```diff
@@ -0,0 +1,395 @@
+        self.on.glauth_snap_ready.emit()
+        cc_content = {"ca-cert": ca_cert}
+        # Get Peer Secrets
+        ldap_relation = self.model.get_relation("glauth")
+        default_bind_dn = ldap_relation.data[self.charm.app]["ldap-default-bind-dn"]
+        ldap_password = ldap_relation.data[self.charm.app]["ldap-password"]
+        ldbd_secret = self.model.get_secret(id=default_bind_dn)
+        lp_secret = self.model.get_secret(id=ldap_password)
+        # Create Secrets
+        cc_secret = self.charm.app.add_secret(cc_content, label="ca-cert")
+        logger.debug("created secret %s", cc_secret)
+        cc_secret.grant(event.relation)
+        ldbd_secret.grant(event.relation)
+        lp_secret.grant(event.relation)
+        event.relation.data[self.charm.app]["ca-cert"] = cc_secret.id
+        event.relation.data[self.charm.app]["ldap-default-bind-dn"] = ldbd_secret.id
+        event.relation.data[self.charm.app]["ldap-password"] = lp_secret.id
+        # Configuration data update
+        ldap_relation = self.model.get_relation("ldap-client")
+        ldap_relation.data[self.charm.app].update(
+            {
+                "basedn": self.model.config["ldap-search-base"],
+                "ldap-uri": ldap_uri,
+            }
+        )
+        self.charm.unit.status = ActiveStatus("glauth ready")
+
+    def load(self) -> str:
+        """Load ca-certificate from glauth snap.
+
+        Returns:
+            The ca certificate content.
+        """
+        cert = "/var/snap/glauth/common/etc/glauth/certs.d/glauth.crt"
+        key = "/var/snap/glauth/common/etc/glauth/keys.d/glauth.key"
+        if not pathlib.Path(cert).exists() and not pathlib.Path(key).exists():
+            # If cert and key do not exist, create both
+            subprocess.run(
+                shlex.split(
+                    f'openssl req -x509 -newkey rsa:4096 -keyout {key} -out {cert} -days 365 -nodes -subj "/CN={self._get_hostname()}"'
+                )
+            )
+        content = open(cert, "r").read()
+        return content
+
+    def set_config(self, config: pathlib.Path, ldap_port: int, api_port: int) -> str:
+        """Set GLAuth config resource. Create default if none found.
+
+        Args:
+            config: Resource config Path object.
+            ldap_port: LDAP port for default config.
+            api_port: API port for default config.
+
+
+        Returns:
+            LDAP URI.
+        """
+        ldap_uri = "ldap"
+        # Create default config with no users if resource glauth.cfg not found
+        if config is None:
+            ldap_uri = ldap_uri + f"://{self._get_hostname()}:{ldap_port}"
+        # Zip file of multiple configs
+        else:
+            with zipfile.ZipFile(config, "r") as zip:
+                zip.extractall("/var/snap/glauth/common/etc/glauth/glauth.d/")
+            ldap_uri = ldap_uri + f"s://{self._get_hostname()}:{ldap_port}"
+        return ldap_uri
```

Comment:
Hmm... I do not believe you need the local variable `ldap_uri` here. You could have it as the following:

```suggestion
        # Create default config with no users if resource glauth.cfg not found
        if config is None:
            ldap_uri = f"ldap://{self._get_hostname()}:{ldap_port}"
        # Zip file of multiple configs
        else:
            with zipfile.ZipFile(config, "r") as zip:
                zip.extractall("/var/snap/glauth/common/etc/glauth/glauth.d/")
            ldap_uri = f"ldap://{self._get_hostname()}:{ldap_port}"
        return ldap_uri
```

Also, you have an extra "s" at the beginning of the f-string in the else branch: `f"s://{self._get_hostname()}:{ldap_port}"`. Is this intentional?

Reply:
The default behavior is to use ldap over ldaps so for the default uri it's `ldap://<hostname>:<port>` and when a custom set of 1 or more configs is supplied it's ldaps ie `ldap://<hostname>:<port>`. Also, maybe I got it mixed up but I thought variables declared in if conditionals and then used outside was not good practice so it's recommended to define them outside the conditional scope.

Reply:
I would think that there would be an additional check to ensure that ldaps is preferred over plaintext ldap. Just checking if the config is None feels like a guess more than a fact that ldaps is configured?

Reply:
Instead of None I will check for ldap or ldaps and have it default to ldap if ldaps is not possible.